# Notebook 2 : Entraînement semi-supervisé

## 1. Préparation des jeux de données

Cette partie charge les données issues du notebook 1, vérifie leur intégrité, sépare le jeu fort du jeu faible, puis découpe le jeu fort en un jeu d'entraînement et un jeu de test stratifié.

Une vérification finale garantit qu'aucune image ne se retrouve dans plusieurs ensembles à la fois.

### 1.1. Imports

Il convient d'importer les packages nécessaires.

In [1]:
# Librairies nécessaires
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split

RANDOM_STATE = 42
EXPECTED_NB_ROWS = 1506

CSV_DIR = Path("../data")


### 1.2. Chargement des données

Chargement du CSV exporté à la fin du notebook 1, contenant les 1506 chemins d'images avec leur label et leur source (jeu fort ou jeu faible).

In [2]:
# Création d’un DataFrame global
df = pd.read_csv(str(CSV_DIR) + "/labels_export.csv", delimiter=";")

### 1.3. Vérification de l'intégrité des données

Deux contrôles rapides avant toute manipulation :
- le nombre total de lignes doit correspondre aux 1506 images attendues
- aucun chemin d'image ne doit apparaître en double dans le CSV

In [3]:
# Vérification du nombre d’éléments
if int(df.shape[0]) != EXPECTED_NB_ROWS:
    print(f"ERREUR : {EXPECTED_NB_ROWS} éléments sont attendus, alors qu'il y a {int(df.shape[0])} éléments extraits.")
else:
    print(f"Il y a bien {EXPECTED_NB_ROWS} éléments dans le DataFrame créé.")

# Vérification des doublons de chemins
duplicates_subset = df.duplicated(subset=['chemin'])

count_duplicates = 0
for index, value in duplicates_subset.items():
    if value == True:
        count_duplicates += 1

if count_duplicates > 0:
    print(f"ALERTE : Il y a {count_duplicates} image(s) en double.")
else:
    print("Aucun doublon détecté")

Il y a bien 1506 éléments dans le DataFrame créé.
Aucun doublon détecté


Le CSV contient bien les 1506 lignes attendues, sans aucun doublon de chemin.

Les données issues du notebook 1 sont fiables pour la suite.

### 1.4. Séparation des sources

Il faut séparer le jeu fort du jeu faible.

In [4]:
# Préparation du DataFrame du jeu fort
df_fort = df[df['source'] == "fort"].copy()

# Préparation du DataFrame du jeu faible
df_faible = df[df['source'] == "faible"].copy()

# Retrait de la colonne "source" sur les deux DataFrames
df_fort.pop('source')
df_faible.pop('source')

print(f"Dimensions du jeu fort : {df_fort.shape}")
print(f"Dimensions du jeu faible : {df_faible.shape}")

Dimensions du jeu fort : (100, 2)
Dimensions du jeu faible : (1406, 2)


Le jeu fort compte 100 images et le jeu faible 1406, conforme à la répartition observée dans le notebook 1.

La séparation par la colonne `source` fonctionne comme attendu.

### 1.5. Séparation des données d'entraînement et de test du jeu fort

Le jeu fort est découpé en un jeu d'entraînement et un jeu de test, avec un split stratifié sur la variable `label` pour préserver la proportion cancer/normal malgré le faible effectif de 100 images.

In [5]:
# Séparation X et y
X = df_fort.drop(columns='label')
y = df_fort['label']

# Split stratifié
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=RANDOM_STATE, stratify=y)

print(f"Dimensions Train : {X_train.shape}")
print(f"Dimensions Test : {X_test.shape}")

Dimensions Train : (80, 1)
Dimensions Test : (20, 1)


Le split stratifié donne 80 images d'entraînement et 20 images de test, dans une proportion 80/20 cohérente avec le paramètre `test_size=0.2`.

La stratification garantit que les deux classes restent représentées dans les mêmes proportions que dans le jeu fort d'origine.

### 1.6. Vérification des fuites de données

Contrôle final garantissant l'absence de chevauchement entre le jeu d'entraînement, le jeu de test et le jeu faible.

C'est la vérification qui matérialise la règle métier de séparation stricte entre données faiblement et fortement labellisées.

In [6]:
# Vérification des fuites de données : aucune image ne doit apparaître dans plus d'un ensemble
chemins_train = set(X_train['chemin'])
chemins_test = set(X_test['chemin'])
chemins_faible = set(df_faible['chemin'])

overlap_train_test = chemins_train & chemins_test
overlap_train_faible = chemins_train & chemins_faible
overlap_test_faible = chemins_test & chemins_faible

if overlap_train_test:
    print(f"FUITE DE DONNÉES : {len(overlap_train_test)} image(s) présentes à la fois dans train et test")
else:
    print("Aucun chevauchement entre train et test")

if overlap_train_faible:
    print(f"FUITE DE DONNÉES : {len(overlap_train_faible)} image(s) présentes à la fois dans train et le jeu faible")
else:
    print("Aucun chevauchement entre train et le jeu faible")

if overlap_test_faible:
    print(f"FUITE DE DONNÉES : {len(overlap_test_faible)} image(s) présentes à la fois dans test et le jeu faible")
else:
    print("Aucun chevauchement entre test et le jeu faible")

Aucun chevauchement entre train et test
Aucun chevauchement entre train et le jeu faible
Aucun chevauchement entre test et le jeu faible


Aucun chevauchement détecté entre train, test et le jeu faible.

La séparation stricte entre données faiblement et fortement labellisées est respectée, et le jeu de test est bien constitué de données jamais vues à l'entraînement.